content/developer/packaging/package-testing-guide.md

```diff
@@ -0,0 +1,133 @@
+
+*Executors* are environments for running the tests. A executor is an abstraction with:
+- A working AOSC OS, with systemd and dbus mostly usable
+- Same architecture with the package
+- autobuild3 with same version installed
+- read ability of the build directory (temporary writes can be implemented by overlay filesystem or simply a copy, so write is not a problem here)
+- an interface for autobuild3 to run commands in it and stop it
+- pipes connected to the autobuild3's stdout and stderr for log output
+- a result report method
+
+Some example executors are:
+- `plain`: the default one, just run it in a subshell, enough for most packages
+- `systemd-run`: resource/time controls, permission limitations
+- `x11`: with a x11 server running and DISPLAY variable set
+- `wayland`: basically the same with x11
+- `qemu`: kernel related tests, namespace operations
+
+Except for `plain` executor, the test specification file is sourced at least twice, once in the host autobuild3's subshell, and once in the executor.
+
+The test *types* defines the pre-defined test script to use. A special "pre-defined" test script called "custom" can invoke your own test script.
+
+## Files
+
+For packages that uses the default test, only `defines` (or other name and location that autobuild3 recognizes) is involved.
+
+For normal testing setup, there must be a `tests/` directory under `autobuild/` that contains all of the test specification file, which is named by the test name, without suffixes.
+
+Test data including self-written scripts, example file for the software to process, or standard outputs or screenshots for comparison should be placed in `autobuild/testdata`. Large files, copyrighted assets, and other controversial files should be hosted elsewhere, keeping the aosc-os-abbs repository clean. Placing files by test name is recommended.
+
+Example:
+
+```
+autobuild/
+    defines
+    tests/
+        screenshot
+        ctest
+    testdata/
+        screenshot/
+            startup.png
+```
+
+## Metadata
+
+All metadata entries should placed in the test specification file, except for `TESTS`.
+
+### `TESTS`
+
+A space separated list of enabled tests in the `defines` file.
+
+The default value is `"default"`, which runs the only test specified in `defines`.
+
+Example:
+```
+TESTS=""                    # disables the default test
+TESTS="default"
+TESTS="screenshot ctest"
+
+TESTS="default ctest"       # incorrect! default should not be used with other tests
+```
+
+### `TESTDESC`
+
+A human readable short description of the test. Must be defined except for the default test.
+
+### `TESTDEP`
+
+A space separated list of dependency for running the test. Defaults to `""`.
+
+Note that this may be only installed inside of the executor, so they may not be available in other tests and other parts of the autobuild3l.
```

Comment:
```suggestion
Note that this may be only installed inside of the executor, so they may not be available in other tests and other parts of the autobuild3.
```